File: Web/src/lib/metrics/goalProjection.ts
/**
 * Goal Projection Calculator
 * 
 * Dynamically calculates projected goal times based on:
 * - Current effective VO2max
 * - Training plan settings (duration, frequency, volume)
 * - Current marathon shape percentage
 * - Race distance
 * 
 * Uses training science principles to estimate VDOT improvement
 * over the plan duration.
 */

import { predictRaceTime, type RaceDistance } from './vdot';

// ============================================
// Types and Interfaces
// ============================================

export interface PlanSettings {
    durationWeeks: number;      // Total plan duration in weeks
    runsPerWeek: number;        // Number of running sessions per week
    weeklyMileageGoal: number;  // Target peak weekly volume in km
    raceDistance: RaceDistance; // Target race distance
    taperWeeks?: number;        // Taper phase length
    peakWeeks?: number;         // Peak phase length
    buildWeeks?: number;        // Build phase length
}

export interface ProjectedGoalResult {
    optimalTime: number;        // Fastest possible (100% shape, full improvement)
    projectedTime: number;      // Realistic based on plan + current shape
    conservativeTime: number;   // Slower estimate (safety margin)
    projectedVdot: number;      // Expected VDOT at race time
    improvementPercent: number; // % improvement from current VDOT
    projectedShape: number;     // Expected shape at race time
    shapeImprovementPercent: number; // % shape improvement from training
}

// ============================================
// Constants
// ============================================

/** Maximum VDOT improvement cap (15%) to prevent unrealistic predictions */
const MAX_IMPROVEMENT_FACTOR = 1.15;

/** Base weekly improvement rate from training duration */
const DURATION_IMPROVEMENT_RATE = 0.008; // 0.8% per 4 weeks

/** Improvement bonus from higher running frequency */
const FREQUENCY_IMPROVEMENT_RATE = 0.02; // 2% per 4 runs/week

/** Improvement bonus from higher weekly volume */
const VOLUME_IMPROVEMENT_RATE = 0.015; // 1.5% per 50km/week

/** Shape impact multipliers by distance (matching runalyze.ts) */
const SHAPE_IMPACT: Record<RaceDistance, number> = {
    '5K': 0.05,     // 5% penalty at 0% shape
    '10K': 0.08,    // 8% penalty at 0% shape
    'HALF': 0.15,   // 15% penalty at 0% shape
    'MARATHON': 0.30, // 30% penalty at 0% shape
};

/** 
 * Shape improvement rate from weekly mileage
 * Based on: Marathon shape improves ~2% per 10km/week increase
 * E.g., going from 30km/week to 50km/week adds ~4% shape
 */
const SHAPE_IMPROVEMENT_PER_10KM = 2.0;

/**
 * Shape improvement rate from training duration
 * Consistent training over time builds endurance base
 * ~1% shape improvement per 4 weeks of training
 */
const SHAPE_IMPROVEMENT_PER_4WEEKS = 1.0;

// ============================================
// Core Functions
// ============================================

/**
 * Calculate the progression coefficient based on training plan parameters
 * 
 * The coefficient represents expected VDOT improvement over the plan duration.
 * Based on training science: ~0.5-1% VO2max improvement per 4 weeks.
 * 
 * @param durationWeeks - Total weeks of training
 * @param runsPerWeek - Number of running sessions per week
 * @param weeklyVolumeKm - Target weekly volume in km
 * @returns Progression factor (1.0 = no improvement, 1.15 = 15% improvement max)
 */
export function calculateProgressionCoefficient(
    durationWeeks: number,
    runsPerWeek: number,
    weeklyVolumeKm: number
): number {
    // Guard against invalid inputs
    if (durationWeeks <= 0) return 1.0;

    // Calculate individual contribution factors
    // Duration: ~0.8% per 4 weeks of training
    const durationContribution = (durationWeeks / 4) * DURATION_IMPROVEMENT_RATE;

    // Frequency: Higher running frequency = better adaptation
    // Normalized around 4 runs/week
    const frequencyContribution = (runsPerWeek / 4) * FREQUENCY_IMPROVEMENT_RATE;

    // Volume: ~1.5% per 50km/week
    // This captures the aerobic development from higher mileage
    const volumeContribution = (weeklyVolumeKm / 50) * VOLUME_IMPROVEMENT_RATE;

    // Total progression factor
    const progressionFactor = 1 + durationContribution + frequencyContribution + volumeContribution;

    // Cap at maximum improvement to prevent unrealistic predictions
    return Math.min(progressionFactor, MAX_IMPROVEMENT_FACTOR);
}

/**
 * Calculate shape penalty for a given race distance
 * 
 * Reuses the same penalty structure as calculateAllRacePredictions in runalyze.ts
 * to ensure consistency across the application.
 * 
 * @param raceDistance - Target race distance
 * @param currentShapePercent - Current marathon shape percentage (0-100+)
 * @returns Penalty factor (0.0 at 100% shape, varies by distance at 0% shape)
 */
export function calculateShapePenalty(
    raceDistance: RaceDistance,
    currentShapePercent: number
): number {
    const shapeImpact = SHAPE_IMPACT[raceDistance] || 0.30;

    // Penalty decreases as shape increases
    // At 100% shape: penalty = 0
    // At 0% shape: penalty = shapeImpact (5-30% depending on distance)
    const penalty = (1 - Math.min(currentShapePercent, 100) / 100) * shapeImpact;

    return penalty;
}

/**
 * Calculate projected goal time based on current fitness and plan settings
 * 
 * This is the main entry point for the projection algorithm.
 * It combines:
 * 1. Current VO2max/VDOT
 * 2. Projected improvement from training plan
 * 3. Shape penalty for the target distance
 * 4. Shape improvement from training mileage
 * 
 * @param currentVO2max - User's current effective VO2max
 * @param planSettings - Training plan configuration
 * @param currentShapePercent - Current marathon shape percentage (default 70)
 * @param currentWeeklyKm - Current average weekly mileage (default: half of target)
 * @returns Projected goal times (optimal, realistic, conservative)
 */
export function calculateProjectedGoalTime(
    currentVO2max: number,
    planSettings: PlanSettings,
    currentShapePercent: number = 70,
    currentWeeklyKm?: number
): ProjectedGoalResult {
    // Default result for invalid inputs
    if (currentVO2max <= 0 || planSettings.durationWeeks <= 0) {
        return {
            optimalTime: 0,
            projectedTime: 0,
            conservativeTime: 0,
            projectedVdot: 0,
            improvementPercent: 0,
            projectedShape: 0,
            shapeImprovementPercent: 0,
        };
    }

    // Step 1: Calculate expected VDOT improvement from training
    const progressionFactor = calculateProgressionCoefficient(
        planSettings.durationWeeks,
        planSettings.runsPerWeek,
        planSettings.weeklyMileageGoal
    );

    // Step 2: Project future VDOT
    const projectedVdot = currentVO2max * progressionFactor;
    const improvementPercent = (progressionFactor - 1) * 100;

    // Step 3: Calculate shape improvement from increased mileage
    // Assume current mileage is half of target if not provided
    const effectiveCurrentKm = currentWeeklyKm ?? (planSettings.weeklyMileageGoal * 0.5);
    const mileageIncrease = Math.max(0, planSettings.weeklyMileageGoal - effectiveCurrentKm);

    // Shape improvement from mileage increase: 2% per 10km/week increase
    const shapeFromMileage = (mileageIncrease / 10) * SHAPE_IMPROVEMENT_PER_10KM;

    // Shape improvement from training duration: 1% per 4 weeks
    const shapeFromDuration = (planSettings.durationWeeks / 4) * SHAPE_IMPROVEMENT_PER_4WEEKS;

    // Total shape improvement (capped so we don't exceed 100%)
    const totalShapeImprovement = shapeFromMileage + shapeFromDuration;
    const projectedShape = Math.min(100, currentShapePercent + totalShapeImprovement);
    const shapeImprovementPercent = projectedShape - currentShapePercent;

    // Step 4: Calculate optimal time (at 100% shape with full VDOT improvement)
    // This represents the theoretical best the athlete can achieve
    const optimalTime = predictRaceTime(projectedVdot, planSettings.raceDistance);

    // Step 5: Calculate projected time (using calculated shape improvement)
    const projectedPenalty = calculateShapePenalty(planSettings.raceDistance, projectedShape);
    const projectedTime = Math.round(optimalTime * (1 + projectedPenalty));

    // Step 6: Calculate conservative time (assumes less improvement)
    // Use current shape with only half the VDOT improvement
    const conservativeVdot = currentVO2max * (1 + (progressionFactor - 1) * 0.5);
    const conservativeBase = predictRaceTime(conservativeVdot, planSettings.raceDistance);
    const conservativePenalty = calculateShapePenalty(planSettings.raceDistance, currentShapePercent);
    const conservativeTime = Math.round(conservativeBase * (1 + conservativePenalty));

    return {
        optimalTime: Math.round(optimalTime),
        projectedTime,
        conservativeTime,
        projectedVdot: Math.round(projectedVdot * 10) / 10,
        improvementPercent: Math.round(improvementPercent * 10) / 10,
        projectedShape: Math.round(projectedShape),
        shapeImprovementPercent: Math.round(shapeImprovementPercent * 10) / 10,
    };
}

/**
 * Calculate duration in weeks from race date
 * 
 * Helper function to calculate plan duration from a race date.
 * 
 * @param raceDate - Target race date
 * @param startDate - Plan start date (defaults to today)
 * @returns Number of weeks until race
 */
export function calculateWeeksUntilRace(
    raceDate: Date,
    startDate: Date = new Date()
): number {
    const msPerWeek = 7 * 24 * 60 * 60 * 1000;
    const diffMs = raceDate.getTime() - startDate.getTime();
    return Math.max(0, Math.floor(diffMs / msPerWeek));
}

File: flutter/lib/core/utils/goal_projection.dart
import 'dart:math';

import 'package:runflow_flutter/core/utils/vdot.dart';
import 'package:runflow_flutter/domain/entities/dashboard_entities.dart';

const _maxImprovementFactor = 1.15;
const _durationImprovementRate = 0.008;
const _frequencyImprovementRate = 0.02;
const _volumeImprovementRate = 0.015;

const _shapeImpact = <RaceType, double>{
  RaceType.fiveK: 0.05,
  RaceType.tenK: 0.08,
  RaceType.halfMarathon: 0.15,
  RaceType.marathon: 0.30,
};

const _shapeImprovementPer10Km = 2.0;
const _shapeImprovementPer4Weeks = 1.0;

double _raceDistanceMeters(RaceType type) {
  switch (type) {
    case RaceType.fiveK:
      return 5000;
    case RaceType.tenK:
      return 10000;
    case RaceType.halfMarathon:
      return 21097.5;
    case RaceType.marathon:
      return 42195;
    case RaceType.fiftyK:
      return 50000;
    case RaceType.fiftyMile:
      return 80467;
    case RaceType.hundredK:
      return 100000;
    case RaceType.hundredMile:
      return 160934;
    case RaceType.twelveHour:
    case RaceType.twentyFourHour:
    case RaceType.backyardUltra:
    case RaceType.customDistance:
    case RaceType.sprintTri:
    case RaceType.olympicTri:
    case RaceType.halfIronman:
    case RaceType.fullIronman:
    case RaceType.customTri:
      return 0;
  }
}

class PlanSettings {
  const PlanSettings({
    required this.durationWeeks,
    required this.runsPerWeek,
    required this.weeklyMileageGoal,
    required this.raceDistance,
  });

  final int durationWeeks;
  final int runsPerWeek;
  final double weeklyMileageGoal;
  final RaceType raceDistance;
}

class ProjectedGoalResult {
  const ProjectedGoalResult({
    required this.optimalTime,
    required this.projectedTime,
    required this.conservativeTime,
    required this.projectedVdot,
    required this.improvementPercent,
    required this.projectedShape,
    required this.shapeImprovementPercent,
  });

  final int optimalTime;
  final int projectedTime;
  final int conservativeTime;
  final double projectedVdot;
  final double improvementPercent;
  final int projectedShape;
  final double shapeImprovementPercent;
}

/// Calculates the expected VDOT improvement ratio based on training parameters.
///
/// Returns a coefficient in the range [1.0, 1.15] representing how much a
/// runner's VDOT is expected to improve over the training plan. A value of
/// 1.0 means no improvement; 1.15 means up to 15% improvement.
///
/// The model uses a linear approximation of the training adaptation curves
/// described in Daniels (2014), *Daniels' Running Formula*, combining three
/// independent contributions:
///
/// - **Duration** ([durationWeeks]): longer plans allow more adaptation.
/// - **Frequency** ([runsPerWeek]): more sessions per week accelerate gains.
/// - **Volume** ([weeklyVolumeKm]): higher weekly mileage builds aerobic base.
///
/// An assertion + clamp guardrail at 1.20 catches unexpected inputs while
/// still returning a safe value in release builds.
///
/// **Parameters:**
/// - [durationWeeks] – plan length in weeks (int, >= 0).
/// - [runsPerWeek]   – scheduled runs per week (int, >= 0).
/// - [weeklyVolumeKm] – target weekly mileage in kilometres (double, >= 0).
///
/// **Output range:** 1.0 (no improvement) … 1.15 (maximum 15% improvement).
///
/// **Scientific basis:** linear approximation of Daniels (2014) training
/// adaptation curves, with empirically-tuned per-contribution rates.
double calculateProgressionCoefficient(
  int durationWeeks,
  int runsPerWeek,
  double weeklyVolumeKm,
) {
  if (durationWeeks <= 0) return 1.0;
  if (runsPerWeek <= 0) return 1.0;

  final durationContribution =
      (durationWeeks / 4) * _durationImprovementRate;
  final frequencyContribution =
      (runsPerWeek / 4) * _frequencyImprovementRate;
  final volumeContribution =
      (weeklyVolumeKm / 50) * _volumeImprovementRate;

  final progressionFactor =
      1 + durationContribution + frequencyContribution + volumeContribution;

  final result = min(progressionFactor, _maxImprovementFactor);

  assert(result >= 1.0 && result <= 1.20,
      'Progression coefficient $result out of expected range [1.0, 1.20]');
  return result.clamp(1.0, 1.20);
}

double calculateShapePenalty(
    RaceType raceDistance, double currentShapePercent) {
  final shapeImpact = _shapeImpact[raceDistance] ?? 0.30;
  return (1 - min(currentShapePercent, 100) / 100) * shapeImpact;
}

ProjectedGoalResult calculateProjectedGoalTime(
  double currentVO2max,
  PlanSettings planSettings, {
  double currentShapePercent = 70,
  double? currentWeeklyKm,
}) {
  if (_raceDistanceMeters(planSettings.raceDistance) == 0) {
    return const ProjectedGoalResult(
      optimalTime: 0,
      projectedTime: 0,
      conservativeTime: 0,
      projectedVdot: 0,
      improvementPercent: 0,
      projectedShape: 0,
      shapeImprovementPercent: 0,
    );
  }

  if (currentVO2max <= 0 || planSettings.durationWeeks <= 0) {
    return const ProjectedGoalResult(
      optimalTime: 0,
      projectedTime: 0,
      conservativeTime: 0,
      projectedVdot: 0,
      improvementPercent: 0,
      projectedShape: 0,
      shapeImprovementPercent: 0,
    );
  }

  final progressionFactor = calculateProgressionCoefficient(
    planSettings.durationWeeks,
    planSettings.runsPerWeek,
    planSettings.weeklyMileageGoal,
  );

  final projectedVdot = currentVO2max * progressionFactor;
  final improvementPercent = (progressionFactor - 1) * 100;

  final effectiveCurrentKm =
      currentWeeklyKm ?? (planSettings.weeklyMileageGoal * 0.5);
  final mileageIncrease =
      max(0.0, planSettings.weeklyMileageGoal - effectiveCurrentKm);

  final shapeFromMileage =
      (mileageIncrease / 10) * _shapeImprovementPer10Km;
  final shapeFromDuration =
      (planSettings.durationWeeks / 4) * _shapeImprovementPer4Weeks;

  final totalShapeImprovement = shapeFromMileage + shapeFromDuration;
  final projectedShape =
      min(100.0, currentShapePercent + totalShapeImprovement);
  final shapeImprovementPercent = projectedShape - currentShapePercent;

  final distanceMeters = _raceDistanceMeters(planSettings.raceDistance);

  final optimalTimeMinutes = estimateTime(projectedVdot, distanceMeters);
  final optimalTimeSeconds = (optimalTimeMinutes * 60).round();

  if (optimalTimeSeconds <= 0) {
    return const ProjectedGoalResult(
      optimalTime: 0,
      projectedTime: 0,
      conservativeTime: 0,
      projectedVdot: 0,
      improvementPercent: 0,
      projectedShape: 0,
      shapeImprovementPercent: 0,
    );
  }

  final projectedPenalty =
      calculateShapePenalty(planSettings.raceDistance, projectedShape);
  final projectedTime =
      (optimalTimeSeconds * (1 + projectedPenalty)).round();

  final conservativeVdot =
      currentVO2max * (1 + (progressionFactor - 1) * 0.5);
  final conservativeBaseMinutes =
      estimateTime(conservativeVdot, distanceMeters);
  final conservativeBaseSeconds = (conservativeBaseMinutes * 60).round();
  final conservativePenalty = calculateShapePenalty(
      planSettings.raceDistance, currentShapePercent);
  final conservativeTime =
      (conservativeBaseSeconds * (1 + conservativePenalty)).round();

  return ProjectedGoalResult(
    optimalTime: optimalTimeSeconds,
    projectedTime: projectedTime,
    conservativeTime: conservativeTime,
    projectedVdot: (projectedVdot * 10).round() / 10.0,
    improvementPercent: (improvementPercent * 10).round() / 10.0,
    projectedShape: projectedShape.round(),
    shapeImprovementPercent:
        (shapeImprovementPercent * 10).round() / 10.0,
  );
}

int estimateTimeForDistance(double vdot, int distanceMeters) {
  final timeMinutes = estimateTime(vdot, distanceMeters.toDouble());
  return (timeMinutes * 60).round();
}

ProjectedGoalResult calculateProjectedGoalTimeForDistance(
  double currentVO2max,
  int durationWeeks,
  int runsPerWeek,
  double? weeklyMileageGoalKm,
  int distanceMeters, {
  double currentShapePercent = 70,
}) {
  if (currentVO2max <= 0 || durationWeeks <= 0 || distanceMeters <= 0) {
    return const ProjectedGoalResult(
      optimalTime: 0,
      projectedTime: 0,
      conservativeTime: 0,
      projectedVdot: 0,
      improvementPercent: 0,
      projectedShape: 0,
      shapeImprovementPercent: 0,
    );
  }

  final progressionFactor = calculateProgressionCoefficient(
    durationWeeks,
    runsPerWeek,
    weeklyMileageGoalKm ?? 0,
  );

  final projectedVdot = currentVO2max * progressionFactor;
  final improvementPercent = (progressionFactor - 1) * 100;

  const shapeImpactBase = 0.30;
  final projectedPenalty =
      (1 - (currentShapePercent + 5).clamp(0, 100) / 100) * shapeImpactBase;

  final optimalTime = estimateTimeForDistance(projectedVdot, distanceMeters);
  final projectedTime = (optimalTime * (1 + projectedPenalty)).round();

  final conservativeVdot =
      currentVO2max * (1 + (progressionFactor - 1) * 0.5);
  final conservativeBase =
      estimateTimeForDistance(conservativeVdot, distanceMeters);
  final conservativePenalty =
      (1 - currentShapePercent.clamp(0, 100) / 100) * shapeImpactBase;
  final conservativeTime =
      (conservativeBase * (1 + conservativePenalty)).round();

  return ProjectedGoalResult(
    optimalTime: optimalTime.round(),
    projectedTime: projectedTime,
    conservativeTime: conservativeTime,
    projectedVdot: (projectedVdot * 10).round() / 10.0,
    improvementPercent: (improvementPercent * 10).round() / 10.0,
    projectedShape: 0,
    shapeImprovementPercent: 0,
  );
}
